core/src/proto/types.rs

```diff
@@ -360,6 +371,146 @@ impl Code {
     }
 }
 
+#[derive(
+    Clone,
+    Debug,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+)]
+pub struct SignatureIndex {
+    pub signature: common::Signature,
+    pub index: u8,
+}
+
+impl SignatureIndex {
+    pub fn from_single_signature(signature: common::Signature) -> Self {
+        Self {
+            signature,
+            index: 0,
+        }
+    }
+
+    pub fn to_vec(&self) -> Vec<Self> {
+        vec![self.clone()]
+    }
+
+    pub fn verify(
+        &self,
+        public_key_index_map: &AccountPublicKeysMap,
+        data: &impl SignableBytes,
+    ) -> std::result::Result<(), VerifySigError> {
+        let public_key =
+            public_key_index_map.get_public_key_from_index(self.index);
+        if let Some(public_key) = public_key {
+            common::SigScheme::verify_signature(
+                &public_key,
+                data,
+                &self.signature,
+            )
+        } else {
+            Err(VerifySigError::MissingData)
+        }
+    }
+}
+
+impl Ord for SignatureIndex {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.index.cmp(&other.index)
```

Comment:
I think if you put the `index` field first in the struct then you can derive the impl as it will be sorted by the index (you might need to add `Ord` to `common::Signature` which is fine)

Reply:
ah, I see. I think we can add it, we already have a manual PartialOrd using the bytes:

```rust
impl Ord for Signature {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.to_bytes().cmp(&other.0.to_bytes())
    }
}
```